keps/sig-api-machinery/3488-cel-admission-control/README.md

```diff
@@ -0,0 +1,1587 @@
+
+Each `ValidatingAdmissionPolicy` resource defines a admission control policy.
+The resource contains the CEL expressions to validate the admission policy and
+declares how the admission policy may be configured for use.
+
+For example:
+
+```yaml
+# Policy definition
+apiVersion: admissionregistration.k8s.io/v1alpha1
+kind: ValidatingAdmissionPolicy
+metadata:
+  name: "validate-xyz.example.com"
+spec:
+  config:
+    group: rules.example.com
+    kind: ReplicaLimit
+    version: v1
+  constraints:
+    rules:
+    - apiGroups:   ["apps"]
+      apiVersions: ["v1"]
+      operations:  ["CREATE", "UPDATE"]
+      resources:   ["deployments"]
+      scope:       "*"
+  validations:
+    - expression: "object.spec.replicas <= config.maxReplicas"
+      # ...other rule related fields here...
+```
+
+The `spec.config` field of the `ValidatingAdmissionPolicy` references the
+"configuration CRD" for this admission policy. For this example, the
+configuration CRD is `ReplicaLimit`.
+
+Note: This is a "Bring Your Own CRD" design. The admission policy definition
+author is responsible for providing the `ReplicaLimit` configuration CRD.
+
+`spec.constraints` specifies that the CEL expressions in
+`validations.expression` have access to the fields of the `v1.Deployment` kind.
+Multiple resources of the same kind may be matched, e.g. `*/scale`.
+
+To configure an admission policy, "policy configurations" of the configuration
+CRD kind are created. For example:
+
+```yaml
+# Policy configurations
+apiVersion: rules.example.com/v1
+kind: ReplicaLimit
+metadata:
+  name: "replica-limit-test.example.com"
+spec:
+  match:
+    namespaceSelectors:
+    - key: environment,
+      operator: In,
+      values: ["test"]
+  maxReplicas: 3
+```
+
+This policy configuration limits deployments to a max of 3 repliacas in all
+namespaces in the test environment.
+
+An admission policy may have multiple configurations. To configure the "prod"
+environment to have a maxReplicas limit of 100, create:
+
+```yaml
+# Policy configurations
+apiVersion: rules.example.com/v1
+kind: ReplicaLimit
+metadata:
+  name: "replica-limit-prod.example.com"
+spec:
+  match:
+    namespaceSelectors:
+    - key: environment,
+      operator: In,
+      values: ["prod"]
+  maxReplicas: 100
+```
+
+This design separates admission policy _definition_ from _configuration_. This
+has a couple advantages:
+
+- Access to, and delegation of, policy configuration is more manageable. In
+  particular, Kubernetes RBAC works well with this design.
+- Without the separation, the next most obvious API shape would be to encode
+  everything into a single resource, which could easily become very large and
+  run into resource size limits.
+
+##### Policy Constraints
+
+Each `ValidatingAdmissionPolicy` resource may optionally set `spec.constraints`
+to constrain the resources it validates. 
+
+- `spec.constraints` limits which resources may be matched configurations of the
```

Comment:
So does this mean the `spec.match` field in the _configuration_ object will further constraint the resources in scope for this policy that the `spec.constraints` field in the `ValidatingAdmissionPolicy` object constraints?

Reply:
Yes. That's a good way to think about it.